public/app/features/dashboard-scene/saving/useSaveDashboard.ts

```diff
@@ -0,0 +1,83 @@
+import { useAsyncFn } from 'react-use';
+
+import { locationUtil } from '@grafana/data';
+import { locationService, reportInteraction } from '@grafana/runtime';
+import { Dashboard } from '@grafana/schema';
+import appEvents from 'app/core/app_events';
+import { useAppNotification } from 'app/core/copy/appNotification';
+import { updateDashboardName } from 'app/core/reducers/navBarTree';
+import { useSaveDashboardMutation } from 'app/features/browse-dashboards/api/browseDashboardsAPI';
+import { SaveDashboardOptions } from 'app/features/dashboard/components/SaveDashboard/types';
+import { useDispatch } from 'app/types';
+import { DashboardSavedEvent } from 'app/types/events';
+
+import { updateDashboardUidLastUsedDatasource } from '../../dashboard/utils/dashboard';
+import { DashboardScene } from '../scene/DashboardScene';
+
+export function useDashboardSave(isCopy = false) {
+  const dispatch = useDispatch();
+  const notifyApp = useAppNotification();
+  const [saveDashboardRtkQuery] = useSaveDashboardMutation();
+
+  const [state, onSaveDashboard] = useAsyncFn(
+    async (scene: DashboardScene, saveModel: Dashboard, options: SaveDashboardOptions) => {
+      {
+        const result = await saveDashboardRtkQuery({
+          dashboard: saveModel,
+          folderUid: options.folderUid,
+          message: options.message,
+          overwrite: options.overwrite,
+          showErrorAlert: false,
+        });
+
+        if ('error' in result) {
+          throw result.error;
+        }
+
+        const resultData = result.data;
+        scene.saveCompleted(saveModel, resultData, options.folderUid);
+
+        // important that these happen before location redirect below
+        appEvents.publish(new DashboardSavedEvent());
+        notifyApp.success('Dashboard saved');
+
+        //Update local storage dashboard to handle things like last used datasource
+        updateDashboardUidLastUsedDatasource(resultData.uid);
+
+        if (isCopy) {
+          reportInteraction('grafana_dashboard_copied', {
+            name: saveModel.title,
+            url: resultData.url,
+          });
+        } else {
+          reportInteraction(`grafana_dashboard_${resultData.uid ? 'saved' : 'created'}`, {
+            name: saveModel.title,
+            url: resultData.url,
+          });
+        }
+
+        const currentLocation = locationService.getLocation();
+        const newUrl = locationUtil.stripBaseFromUrl(resultData.url);
+
+        if (newUrl !== currentLocation.pathname) {
+          setTimeout(() => locationService.replace({ pathname: newUrl, search: currentLocation.search }));
```

Comment:
I think we should push rather than replace. This breaks navigating back to the source dashboard.

Also - what do you think default behavior that would leave you in the source dashboard when you click Save as? Thinking about how it works in most software, when you Save as, a new copy is created but you are not taken to it. I also wonder if maybe we should rename `Save as` to `Duplicate` or `Make a copy`?

Reply:
Alternatively when Saving as, we could have a checkbox the when selected the user would be taken to the newly created copy?

Reply:
> Also - what do you think default behavior that would leave you in the source dashboard when you click Save as? Thinking about how it works in most software, when you Save as, a new copy is created but you are not taken to it. I also wonder if maybe we should rename Save as to Duplicate or Make a copy?

That feels very wrong, when I use save as google drive opens the new copy. VS code also opens the saved as copy. What we could do if we want is to open save as copy in a new tab. But it feels bad also, let's say you have made some changes to a dashboard and want to save it as a copy. We save it as copy and open copy in a new tab. The original tab would still have those changes and if youy went back there you would have unsaved changes warning when you leave it which would be confusing as you have already saved it (but to a copy). 

I think I like Save as the most, it's how vs code and most file editors name it and match those behaviors the most.  If we had "auto save" like google drive has then "Make a copy" would make sense as well 

@dprokop yea, if it's a copy/new we should do push, but if it's just a rename it should be a replace

Reply:
ok, let's make a `push` then :)